### ISSUE TITLE
Error: No Such Key

### ISSUE THREAD

**Issue by cdenneen** [logstash-plugins/logstash-input-s3#59]
Running Logstash 1.5.4 with logstash-input-s3-1.0.0

```
{:timestamp=>"2015-10-07T06:40:20.333000-0400", :message=>"A plugin had an unrecoverable error. Will restart this plugin.\n  Plugin: <LogStash::Inputs::S3 bucket=>\"my.cloudtrail.bucket1\", prefix=>\"AWSLogs\", access_key_id=>\"MYKEY\", secret_access_key=>\"FOOBAR\", delete=>false, interval=>900, region=>\"us-east-1\", sincedb_path=>\"/var/log/logstash/my.cloudtrail.bucket1.sincedb\", type=>\"cloudtrail\", codec=><LogStash::Codecs::CloudTrail spool_size=>50>, debug=>false, use_ssl=>true, temporary_directory=>\"/var/lib/logstash/logstash\">\n  Error: No Such Key", :level=>:error}
{:timestamp=>"2015-10-07T06:40:51.024000-0400", :message=>"A plugin had an unrecoverable error. Will restart this plugin.\n  Plugin: <LogStash::Inputs::S3 bucket=>\"my.cloudtrail.bucket1\", prefix=>\"AWSLogs\", access_key_id=>\"MYKEY\", secret_access_key=>\"FOOBAR\", delete=>false, interval=>900, region=>\"us-east-1\", sincedb_path=>\"/var/log/logstash/my.cloudtrail.bucket1.sincedb\", type=>\"cloudtrail\", codec=><LogStash::Codecs::CloudTrail spool_size=>50>, debug=>false, use_ssl=>true, temporary_directory=>\"/var/lib/logstash/logstash\">\n  Error: No Such Key", :level=>:error}
{:timestamp=>"2015-10-07T06:41:23.678000-0400", :message=>"A plugin had an unrecoverable error. Will restart this plugin.\n  Plugin: <LogStash::Inputs::S3 bucket=>\"my.cloudtrail.bucket1\", prefix=>\"AWSLogs\", access_key_id=>\"MYKEY\", secret_access_key=>\"FOOBAR\", delete=>false, interval=>900, region=>\"us-east-1\", sincedb_path=>\"/var/log/logstash/my.cloudtrail.bucket1.sincedb\", type=>\"cloudtrail\", codec=><LogStash::Codecs::CloudTrail spool_size=>50>, debug=>false, use_ssl=>true, temporary_directory=>\"/var/lib/logstash/logstash\">\n  Error: No Such Key", :level=>:error}
```

@ph maybe this can use your monkey fix for sqs input to restart?
Seems to work for periods of time but then threads die.
Also what seems odd is I have this running for few accounts and once I get this error I get no more "cloudtrail" events so it's like all the s3 inputs threads die.

**Comment by gunzy83** [comment]
I think I may have come across the same issue.

While testing Logstash 2.0.0 to ingest a copy of about 5 days of our AWS ELB logs from production I had two issues:

1. Slow startup time (due to the plugin reading every file in the bucket (approx 1200 files per day)
2. Plugin would ingest 1 log file at a time, crash and restart with the following error (removed environment specific names):

```
A plugin had an unrecoverable error. Will restart this plugin.
  Plugin: <LogStash::Inputs::S3 type=>"elb", bucket=>"elb-test-log-processing", region=>"ap-southeast-2", sincedb_path=>"/data/app/logstash-2.0.0/.sincedb", interval=>10, backup_to_bucket=>"elb_test_backup", delete=>true, codec=><LogStash::Codecs::Plain charset=>"UTF-8">, use_ssl=>true, temporary_directory=>"/tmp/logstash">                                                                                
  Error: No Such Key {:level=>:error}
```

This resulted in a very low ingestion rate that would never keep up with the production logs.

Config:
```
input {
  s3 {
    type => "elb"
    bucket => "elb-test-log-processing"
    backup_to_bucket=>"elb_test_backup"
    region => "ap-southeast-2"
    sincedb_path => "/data/app/logstash-2.0.0/.sincedb"
    interval => "10"
    delete => "true"
  }
}
```

I noticed that setting delete = "false" seemed to fix the issue and backup_to_bucket would work correctly as well. I did want deletes though so I traced it down to line [327](https://github.com/logstash-plugins/logstash-input-s3/blob/master/lib/logstash/inputs/s3.rb#L327) in s3.rb where it is trying to get the last modified date on an object that no longer exists.

Fix is here is you want to try it out: [https://github.com/gunzy83/logstash-input-s3/commit/09e1700b0c6ff9139b00a542f4e518d1ee261876](https://github.com/gunzy83/logstash-input-s3/commit/09e1700b0c6ff9139b00a542f4e518d1ee261876).

I will write a test for this if required and put in a PR soon.

**Comment by gunzy83** [comment]
This looks to be solved in Logstash 2.2.x.

**Comment by vanga** [comment]
I am using logstash-2.2.1
I still see this issue
```
  s3 {
    bucket => "xxx"
    interval => 1800 # seconds
    prefix => "xyz"
    region => "us-west-1"
    type => "cloudfront-access-logs"
    codec => "plain"
    sincedb_path => "/root/.sincedb/prod-xxx"
    backup_to_bucket => "cf-logs-backup"
    delete => true
  }
```
I am not sure how to confirm if the changes are part of the release.

**Comment by Jeiwan** [comment]
I also experience the issue, although it's quite random: sometime it fails, sometimes it doesn't. Restarting Logstash (without changing AWS credentials) solves the problem.
Logstash 2.3.3, the config is very basic:

```
input {
  s3 {
    bucket => "bucket"
    access_key_id => "key"
    secret_access_key => "secret"
  }
}
```

**Comment by BigDataEngineer** [comment]
Hi, 

we continue to see the errors in our set up.

{:timestamp=>"2016-07-23T14:32:18.447000-0700", :message=>"A plugin had an unrecoverable error. Will restart this plugin.\n  Plugin: <LogStash::Inputs::S3 region=>\"us-west-2\", bucket=>\"esxbectransactionsinput\", prefix=>\"4-slc4b01c-9b20-GQC/\", codec=><LogStash::Codecs::JSONLines charset=>\"UTF-8\", delimiter=>\"\\n\">, delete=>true, interval=>10, sincedb_path=>\"/dev/null\", temporary_directory=>\"/data/logstash/s3_to_logstash/GQC\", use_ssl=>true>\n  Error: No Such Key", :level=>:error}


logstash version 
root@Importer-3a:/opt/logstash/bin# ./logstash --version
logstash 2.3.1

I believe this change is already a part of 2.3.1(see below), still we are getting the errors randomly and logstash is very slow in reading the files from s3 bucket. Can you please advice what is a possible solution?
We have tried restarting logstash and or rebooting the machine itself , but every time there is a big amount of files to process , we see slowness or crash of logstash with this error in logstash.log files.

Thanks!!
`/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-s3-2.0.6/lib/logstash/inputs/s3.rb

def process_log(queue, key)
    object = @s3bucket.objects[key]

    filename = File.join(temporary_directory, File.basename(key))

    if download_remote_file(object, filename)
      if process_local_log(queue, filename)
        lastmod = object.last_modified
        backup_to_bucket(object, key)
        backup_to_dir(filename)
        delete_file_from_bucket(object)
        FileUtils.remove_entry_secure(filename, true)
        sincedb.write(lastmod)
      end
    else
      FileUtils.remove_entry_secure(filename, true)
    end
  end


`

**Comment by shronny** [comment]
Hi, 
Any news regarding this issue? its happening all the time...
im using version 2.3.4 with the following configuration:

`input {`
        `s3 {`
        `access_key_id => "key"`
        `secret_access_key => "access"`
        `bucket => "bucket"`
        `region => "us-east-1"`
        `sincedb_path => "/tmp/s3.sincedb"`
        `}`
`}`

Resulting this error:
`  Error: No Such Key
  Exception: AWS::S3::Errors::NoSuchKey
  Stack: /opt/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-v1-1.66.0/lib/aws/core/client.rb:375:in `return_or_raise'`

**Comment by waynetaylor** [comment]
+1 to this. I get it even with a small s3 bucket

 wtaylor@wtaylor-mbp  ~/Downloads/logstash-2.2.2/bin  ./logstash -f s3example.conf
Settings: Default pipeline workers: 8
Logstash startup completed
A plugin had an unrecoverable error. Will restart this plugin.
  Plugin: <LogStash::Inputs::S3 bucket=>"gogo-billing-mbb-normalizer-prods/logstash", aws_credentials_file=>"/Users/wtaylor/Downloads/logstash-2.2.2/bin/s3key.txt", delete=>false, codec=><LogStash::Codecs::Plain charset=>"UTF-8">, region=>"us-east-1", use_ssl=>true, interval=>60, temporary_directory=>"/var/folders/zy/1swsp3vd31g9qrcvbsyct8h1zr77l6/T/logstash">
  Error: No Such Key {:level=>:error}

**Comment by mauriziod** [comment]
After a while I still can see this issue happening using:
* logstash 5.3.0-1
* debian 8.7
* java 1.8.0_121

**Comment by robbavey** [comment]
This should have been fixed by #144
